man/fspscan.1
.TH FSPSCAN 1 "Jan 2005" FSP

.SH NAME
fspscan is a small tool to scan host for a running fsp server

.SH SYNOPSIS
.B fspscan
[-t][-o file][-r retries][-l local port][-i ip address][-p starting port][-h][-e endport]

.SH DESCRIPTION
fspscan is small tool to scan for a running fsp server on a remote host. When it  detects a running server it will display
the server version and the port.

.SH ARGUMENTS
.LP
.B -t
-- trace (on|off), default is off
.LP
.B -o
-- file, define a file for the output, default is stdout
.LP
.B -r
-- retries, define the number of probes bevor we skip this port and continue with the next one, default is 3
.LP
.B -l
-- local port, define the local port number we use for scaning, default is 9191
.LP
.B -i
-- ip address, define the ip address or hostname of the host you would like to scan
.LP
.B -p
-- starting port, define the port number where we start to scan, default is 21
.LP
.B -e
-- ending port, define the port number where we stop to scan, default is 65535
.LP
.B -h
-- help, print out a short help and the default values

.SH BUGS
fspscan is not fastest scanning tool on the earth. This program is
known in the wild as FSP exploit. Detection can be avoided by
disabling CC_VERSION command on server side.
.SH HISTORY
This program was written in late 1992 by Cuda and quickly becomes must have
tool for every FSP fan. In September 2003
it was adapted to modern C compilers and added to FSP suite by Radim
Kolar. Because most value of this program is purely historic, fixes
and feature enhancements
of this program are kept at minimal amount.
.SH "SEE ALSO"
.PD
fcatcmd(1), fcdcmd(1), fgetcmd(1), fgrabcmd(1), flscmd(1), fmkdir(1),
fprocmd(1), fput(1), frmcmd(1), frmdircmd(1), fver(1), fducmd(1),
fhostcmd(1), fspd(1), fstatcmd(1), fsp_prof(5), ffindcmd(1), fsp_env(7)


ChangeLog
Version 2.8.1b27 - 9 Sep 2014
	fhostcmd: Fixed output of FSP_TRACE variable for Bourne style shells
	correctly initialize to zero client packet sent statistic counter
	note default timeout values in fsp_env man page
	setting environment value FSP_TIMEOUT to <=0 uses 4byte maxint value
	note default fspd max packet size in fsp_env man page
	fix build on systems with utmpx.h API (FreeBSD 9)
	fspscan: do not shadow host variable
	fspscan: fix clang warning about empty for cycle body by placing
	  semicolon on separate line
	FAQ: Cleaned lot of SGML markup errors
	Do not use sgmlformat, call jade directly
	correctly zero entire sockaddr_in structure during name resolving,
	  error found by Clang.
	allow docbook dsssl stylesheets to be specified at scons command line
	always validate path in SETPRO and MKDIR even if packet is resent and
	  no action will be taken. This is needed to avoid uninitialised
	  pointer access during reply. Found by Clang.

Version 2.8.1b26 - 1 Oct 2009
	fver now reports if server can process extra data in input packets.
	  fspclient did that for a long time. This field is used by fspd
	  2.8.1 beta 21 or newer.
	Server version string sent as response to CC_VERSION command have
	  no longer \n appended to it. Old FSP used it but fver client
	  command can handle server strings without \n for quite a long
	  time, so i decided to remove it. It is bad idea to terminate
	  version strings by \n while it is not required by FSP protocol.
	glob() function used in FSP clients renamed to bsdglob because it
	  clashed with glob() function in newer Linux glibc resulting in
          segfaults.
	non continuous upload detection fixed. Now detection works as follows:
	  if you are trying to upload past end of file, fspd will report that
	  you are trying to create hole in file.
	  if you are seeking backwards more than 1 packet size fspd will
	  report noncontinuous upload. Most likely port from old unfinished
	  download was reused.
	We can now reupload last block which can happen if fspd server is
	  restarted and reply to upload command is lost on way from server
	  to client.
	Check for overwriting uploaded files in fspd was reversed by mistake.
	  Now if you want to overwrite file you must have permission for
	  deleting and adding files or be directory owner.
	Turning on debug in SCons turns on maintainer mode also.
	Upload starting at zero position are now always allowed no matter if
	  source port was reused or not. Old code did that but not over
	  server restarts.
	INFO file updated with fresh information
	fsetupcmd and fhostcmd now reject port number > 65535
	fsetupcmd - code cleaned a bit. Error message is now
	  fsetup: No host given! instead of fhost:
	fsetupcmd got -l option. It lists all known fsp servers collected
	  from .fspsites files.
	fpwd command from setup.sh now prints site name instead of ip
	  address if possible.
	SCons now check if CC compiler works. We need SCons 1.1 or greater
	  for building FSP now.
	Large file support on Linux is now correctly detected.
	Error codes added to CC_ERR command in PROTOCOL. It was long wished
	  feature from ISV. Now there is standard way how to pass error codes
	  from server to client. No such codes are defined yet, but you can
	  use custom error codes in range 0xF000 - 0xFFFF. because no such
	  codes are defined fspd do not sends them in error message, but it
	  now correctly zeroes bb_pos field in CC_ERR reply to indicate
	  that extended error code status is not attached to message. Old
	  fspd versions left this field untouched - it will contain request
	  bb_pos in most, but not all cases. Because we indicate size of
	  extra data to be received, this extended error handling is
	  compatible with "plain FSP over serial" standard.
	Code portability improved. Now we don't need to check sizes of
	  predefined types because we are now using ISO C types. Only thing
	  what we need to know is if largefiles are supported or not.
	  If SCons cant run test program for sizing off_t for any reason,
	  it defaults to 4 (no largefiles, 32-bit system).

Version 2.8.1b25 - 10 Sep 2009
	GNU autotools build system was entirely replaced by SCons
	corrected misleading comment in client/lib.c
	default packetsize changed to do best fit into maximum 1500 MTU
	  Ethernet frame = 1460. This is needed for speedup the booting of
	  Etherboot clients
	PROTOCOL: when computing checksum declare *t as unsigned char * type
	changed typo setguid -> setgid in fspd.conf
	fspd.conf: added note that only numeric uid values works
	fspd.conf: added more information about packetsize
	fspd.1: updated information about packetsize
	fspd.conf: repaired some spelling errors by ispell
	MTU changed from 1500 to 1492 for better operation on pppOE networks
	Max payload changed from 4096 to 4380 bytes (3 packets)
	fspd: command line argument now overrides config file
	updated year in copyright
	fsetup command added to setup.sh
	fsetup command added to setup.csh
	fcd: remove trailing \r\n from readme file
	fcd: do not print empty line after directory protection
	fsetupcmd: don't crash if called with numeric address
	fspscan: committed fix by Yang Hong
	SCONS: turning on maintainer-mode links against efence (if possible)
	  same thing like ./configure do
	fix compilation on systems without large files
	cygwin support disabled. we cant detect it from SCons yet
	frmdir works now. It didn't worked due to adding new access rights
	  DIR_NOLIST and DIR_NOGET
	MACHINES updated. Builds confirmed on Ubuntu 8.04.3,
	  FreeBSD 6.4 and FreeBSD 7.2
	fsetupcmd added to fspmerge client
	SCons can now build SGML FSP FAQ if sgmlfmt is found during build
	  process
        fspmerge now prints short help text if it is invoked as fspmerge
	FILES is up to date now

Version 2.8.1b24 - 19 Feb 2005
	use getopt from stdio.h instead of getopt.h in fspscan.c
	  fspscan now compiles on AIX and other OS without glibc (hoaxter)

contrib/fspscan.c
/*
This is an FSP port scanner that I wrote in late '92.  It's not very fast,
but that was a limit of the original fver code (and of current versions
of fver).  Just type "fspscan -h" to get help on how it's used.
My appologies for the nasty code...I just wanted something that worked.

 -Cuda
*/
/*
I have ported it in September 2003 to modern C and added to FSP software
suite. Don't blame me for the code, i have not touched it.

 -Radim
*/
/*********************************************************************\
*  Copyright (c) 1991 by Wen-King Su (wen-king@vlsi.cs.caltech.edu)   *
*                                                                     *
*  You may copy or modify this file in any manner you wish, provided  *
*  that this notice is always included, and that you hold the author  *
*  harmless for any loss or damage resulting from the installation or *
*  use of this software.                                              *
\*********************************************************************/


/****************************************************************
 * FindPort -- hack of fver to find valid ports on a FSP server *
 * by Cuda.  Yes, it's ugly, but it works for me.               *
 * Type "fspscan -h" to see how it is used.                     *
 ****************************************************************/

#include "fspscan.h"
#include <netdb.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

char *host = NULL;
char *outputfile = NULL;
/* Default values */
int  localport = 9191;
int  remoteport = 21;
int  endport = 65535;
int  retries = 3;
int  client_trace = 0;

static int myfd;
static struct sockaddr_in server_addr;
static unsigned short myseq = 0;
static unsigned short key;
int client_intr_state = 0;

static struct sockaddr_in INET_ZERO;

#define DSIZE (sizeof(int)*8)
#define SAVE(A) { int sav; sav = errno; A; errno = sav; }

static void printhelp(int argc,char **argv)
{
    fprintf(stderr,"%s - an fsp port scanner\n",argv[0]);
    fprintf(stderr,"Usage: %s [-h]\n",argv[0]);
    fprintf(stderr,"       (prints this help screen)\n");
    fprintf(stderr,"       %s [-t][-o file][-r retries][-l local port][-i ip address][-p starting port][-e ending port]\n\n",argv[0]);
    fprintf(stderr,"Defaults: (-o) file = stdout\n");
    fprintf(stderr,"          (-l) local port = 9191\n");
    fprintf(stderr,"          (-p) starting port = 21\n");
    fprintf(stderr,"          (-e) ending port = 65535\n");
    fprintf(stderr,"          (-r) retries = 3\n");
    fprintf(stderr,"          (-t) trace = off\n");
    fprintf(stderr,"          (-i) host-ip = none (host MUST be specified)\n");
    exit(0);
}

static void fdclose(void)
{
    close(myfd);
}

static void client_done(void)
{
    (void) client_interact(CC_BYE,0L,0,NULLP,0,NULLP);
}

int main(int argc,char **argv)
{
    int  i;
    UBUF *ub;
    FILE *logfile;

    while ((i = getopt(argc, argv, "htp:o:r:l:i:e:")) != EOF) {
	switch(i) {
	case 'h':
	    printhelp(argc, argv);
	    break;
        case 't':
	    client_trace=1;
	    break;
        case 'r':
	    retries=atoi(optarg);
	    break;
	case 'p':
	    remoteport=atoi(optarg);
	    break;
	case 'e':
	    endport=atoi(optarg);
	    break;
	case 'l':
	    localport=atoi(optarg);
	    break;
	case 'i':
	    host = (char *) malloc(strlen(optarg)+1);
	    strcpy(host, optarg);
	    break;
        case 'o':
	    outputfile = (char *) malloc(strlen(optarg)+1);
	    strcpy(outputfile, optarg);
	    break;
	case '?':
	    printhelp(argc, argv);
	    break;
	default:
	    fprintf(stderr,"This error should not happen\n");
	}
    }

    if (host == NULL) {
        fprintf(stderr, "host/ip not specified, unable to continue\n");
	printhelp(argc, argv);
	exit(1);
    }

    if (outputfile == NULL) {
	logfile=stdout;
    } else {
        logfile=fopen(outputfile,"a+");
    }

    fprintf(logfile,"Scanning %s\n",host);

    for ( ; remoteport<=endport; ++remoteport) {
        init_client(host,remoteport,localport);
        signal(SIGINT,client_intr);

        ub = client_interact(CC_VERSION,0L, 0,NULLP, 0,NULLP);
	if (ub)
	    /* success! */
            fprintf(logfile,"Found FSP ver: %s on port %0d\n",
			    ub->buf,remoteport);
        else
	    fprintf(logfile,"%0d...nada\n",remoteport);
        fflush(logfile);
        fdclose();
    }
    return 0;
}


UBUF *client_interact(cmd,pos,l1,p1,l2,p2)
    unsigned cmd, l1, l2;
    unsigned long pos;
    unsigned char *p1, *p2;
{
    struct sockaddr_in from;
    UBUF sbuf;
    static UBUF rbuf;
    unsigned char *s, *t, *d;
    unsigned u, n, sum, mlen;
    fd_set mask;
    int retval, retry;
    socklen_t bytes;

    sbuf.cmd = cmd;
    sbuf.len = htons(l1);
    sbuf.pos = htonl(pos);

    client_intr_state = 1;
    FD_ZERO(&mask);

    for(u = l1, d = (unsigned char *) sbuf.buf; u--; *d++ = *p1++);
    for(u = l2				      ; u--; *d++ = *p2++);
    mlen = d - (unsigned char *) &sbuf;

    for(retry = 0; retry < retries; retry++) {
	sbuf.key = key;
	sbuf.seq = myseq;
	sbuf.sum = 0;

	for(t = (unsigned char *) &sbuf, sum = n = mlen; n--; sum += *t++);
	sbuf.sum = sum + (sum >> 8);
	if(client_trace && retry)
	    write(2,"R",1);

	if(sendto(myfd,&sbuf,mlen,0,(struct sockaddr*)&server_addr,sizeof(server_addr)) == -1)
						{ perror("sendto"); exit(1); }
	FD_SET(myfd,&mask);

	while(1) {
	    retval = _x_select(&mask, 3000L);

	    if((retval == -1) && (errno == EINTR)) continue;

	    if(retval == 1) {  /* an incoming message is waiting */
		bytes = sizeof(from);
		if((bytes = recvfrom(myfd,(char*)&rbuf,sizeof(rbuf),0,
	           (struct sockaddr*)&from,&bytes)) < UBUF_HSIZE) continue;

		s = (unsigned char *) &rbuf;
		d = s + bytes;
		u = rbuf.sum; rbuf.sum = 0;
		for(t = s, sum = 0; t < d; sum += *t++);
		sum = (sum + (sum >> 8)) & 0xff;
		if(sum != u) continue;  /* wrong check sum */

		rbuf.len = htons(rbuf.len);
		rbuf.pos = htonl(rbuf.pos);

		if(rbuf.seq 	      != myseq) continue;  /* wrong seq # */
		if(rbuf.len+UBUF_HSIZE > bytes) continue;  /* truncated.  */

		myseq++; key = rbuf.key;	/* update seq and keys.   */

		if(client_intr_state == 2) { client_done(); exit(1); }

		return(&rbuf);

	    } else break;   /* go back to re-transmit buffer again */
	}
    }
    /* return NULL if there have been too many retries - ffindport hack */
    return(NULL);
}


void init_client(char *myhost,int port,int myport)
{
    if((myfd = _x_udp(&myport)) == -1)
		{ perror("socket open"); exit(1); }

    if(_x_adr(myhost,port,&server_addr) == -1)
		{ perror("server addr"); exit(1); }
}


void client_intr(int sig)
{
    switch(client_intr_state) {
        case 0: exit(2);
        case 1: client_intr_state = 2; break;
        case 2: exit(3);
    }
}


int _x_udp(int *port)
{
    int f, zz;
    struct sockaddr_in me ;
    struct sockaddr_in sin;
    socklen_t len;

    me = sin = INET_ZERO;

    me.sin_port = htons((unsigned short) *port);
    me.sin_family = AF_INET;

    if((f=socket(AF_INET,SOCK_DGRAM,0)) == -1) return(-1);

    if( setsockopt(f,SOL_SOCKET,SO_REUSEADDR,&zz,sizeof(zz)) < 0 ||
        bind(f,(struct sockaddr *) &me,(len = sizeof(me))) < 0 ||
        getsockname(f,(struct sockaddr*)&sin,&len) < 0)
                                { SAVE(((void) close(f))); return(-1); }
    if(!*port) *port = ntohs((unsigned short) sin.sin_port); return(f);
}

int _x_adr(char *xhost,int port,struct sockaddr_in *his)
{
    char myhost[128];
    struct hostent *H;
    int    i;
    char *s, *d;

    *his = INET_ZERO;
    if(!xhost) (void) gethostname(xhost = myhost,sizeof(myhost));

    if((his->sin_addr.s_addr = inet_addr(xhost)) != INADDR_NONE) {
	his->sin_family = AF_INET;
    } else
    if((H = gethostbyname(xhost))) {
	for(s = (char *)H->h_addr,
	    d = (char *)&his->sin_addr,
	    i = H->h_length; i--; *d++ = *s++)
	      ;
        his->sin_family = H->h_addrtype;
    } else return(-1);
    his->sin_port = htons((unsigned short) port);

    return(0);
}

int _x_select(fd_set *rf, long tt)       /* tt is in unit of ms */
{
    struct timeval timeout;

    if(tt != -1)
    {
        timeout.tv_sec  =  tt / 1000;
        timeout.tv_usec = (tt % 1000)*1000;
        return(select(DSIZE, rf, NULL, NULL, &timeout));
    }

    return(select(DSIZE, rf, NULL, NULL, (struct timeval *) 0));
}
